sys/ztimer/periph_ptp.c

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright (C)    2018 Kaspar Schleiser <kaspar@schleiser.de>
+                    2021 Otto-von-Guericke-Universität Magdeburg
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser General
+ * Public License v2.1. See the file LICENSE in the top level directory for more
+ * details.
+ */
+
+/**
+ * @ingroup     sys_ztimer_periph_ptp
+ * @{
+ *
+ * @file
+ * @brief       ztimer periph/ptp implementation
+ *
+ * @author      Jana Eisoldt <jana.eisoldt@ovgu.de>
+ *
+ * @}
+ */
+#include "assert.h"
+
+#include "irq.h"
+#include "periph/ptp.h"
+#include "ztimer/periph_ptp.h"
+
+#define ENABLE_DEBUG 0
+#include "debug.h"
+
+static ztimer_clock_t *clock_timer;
+
+void ptp_timer_cb(void)
+{
+    ztimer_handler(clock_timer);
+}
+
+static void _ztimer_periph_ptp_set(ztimer_clock_t *clock, uint32_t val)
+{
+    (void)clock;
+
+    unsigned state = irq_disable();
+
+    clock_timer = clock;
+    ptp_timer_set_u64(val);
```

Comment:
this needs to be underflow safe, if the underlying function is not! e.g., if setting 0, this needs to set a minimal value that will still trigger in the future.

Reply:
This shouldn't be an issue for PTP, but it did make me realize that this will not work this easy: The PTP API assumes that the 64 bit value will never overflow. Hence, if one sets a timeout in the past, it will result in an IRQ right away. (A 64 bit period is just too long to be sensible. And even the hardware that I looked at implemented it this way, that the IRQ is triggered when the current value is greater than or equal to the target value. In fact, it is difficult to implement it differently in hardware: The actual clock frequency will be less than 1 GHz on all MCUs for quite some time, so the nanosecond counter (which is required by PTP as resolution) will always step more than one tick per clock cycle. Hence, only triggering an IRQ once the value is identical to the target value would likely result in the counter stepping over the target value without perfectly hitting it.)

However, if this is tested with a synchronized PTP clock, this will not work as it is: Internally, the time will already greater than `UINT32_MAX`, so this will always trigger an IRQ right away. The best would actually be to allow for 64 bit hardware clocks in ztimer. A quick hack would be to set the time like this:

```C
    uint64_t now = ptp_clock_read_u64();
    uint64_t target = (now & 0xffffffff00000000) | val;
    if (target < now) {
        target += 1ULL << 32;
    }
    ptp_timer_set_u64(target);
```

Reply:
Hm, for ztimer, this behavior is perfect. ztimer only ever sets relative values. I guess that's what `ptp_timer_set_u64()` expects? Otherwise there needs to be a read()/add()/set().

Reply:
(with "trigger in the future" I meant "trigger right away or in the future")

Reply:
> I guess that's what `ptp_timer_set_u64()` expects?

No, that is absolute. Looking at the other APIs, it would have made damn sense to name that `set_absolute()`. But I have another API change in the pipe anyway, so I could just also add `set_absolute()` and make this relative.

It is good that nobody uses the API so far (hopefully!)

Reply:
@Jnae so the conclusion is: the code here is good as it is, but you need to wait for me fixing the API.

Reply:
Wait, [`ptp_timer_set_u64()`](https://api.riot-os.org/group__drivers__periph__ptp.html#ga03812aa5adece70e82398147fce1d517) does expect relative times and there even is [`ptp_timer_set_absolute()`](https://api.riot-os.org/group__drivers__periph__ptp.html#ga7aca6b37425dc052ccf12a4bdf6b054a).

Why I'm always the person with the least knowledge about the code I have written?

So, conclusion is: The code is fine and this should work.

Reply:
> Why I'm always the person with the least knowledge about the code I have written?

that's why we team up. I'm usually the person with the least knowledge about the code I have written.

Reply:
Also: The test in `tests/periph_ptp_timer` includes testing for correct behavior with small relative sets (including a relative set of zero).

Reply:
> Wait, [`ptp_timer_set_u64()`](https://api.riot-os.org/group__drivers__periph__ptp.html#ga03812aa5adece70e82398147fce1d517) does expect relative times and there even is [`ptp_timer_set_absolute()`](https://api.riot-os.org/group__drivers__periph__ptp.html#ga7aca6b37425dc052ccf12a4bdf6b054a).

Great, I was already doubting my skills of testing. Sorry that I didn't read your message until now, otherwise I would have told you.